### PR TITLE
renew token

### PR DESCRIPTION
**Pull request by roboll**
Adds a flag, `-renew-token`, which starts a goroutine to renew the token according to its ttl. It renews the token each `tokenttl / 2` period, and in case of failure reschedules in `remaining ttl / 2`.

Required updating to a newer version of vault lib, so there's some godep changes in here too. 0632ce8 has the actual changes.

### CODE REVIEW

File: vault.go [hunk 1]
```diff
@@ -481,6 +481,39 @@ func newVaultClient(opts *config) (*api.Client, error) {
 	// step: set the token for the client
 	client.SetToken(token)
 
+	if opts.vaultRenewToken {
+		tokeninfo, err := client.Auth().Token().LookupSelf()
```
Comment (gambol99) [inline vault.go]: out of interest is this dependent on a vault version?

File: vault.go [hunk 2]
```diff
@@ -481,6 +481,39 @@ func newVaultClient(opts *config) (*api.Client, error) {
...
+		glog.Infof("token ttl is %v", tokenttl)
+		renewPeriod := tokenttl / 2
+		go func() {
+			for {
+				glog.Infof("scheduling token renew in %v", renewPeriod)
+				<-time.After(renewPeriod)
+				glog.Infof("attempting to renew token")
+				newtokeninfo, err := client.Auth().Token().RenewSelf(0)
+				if err != nil {
+					renewPeriod = renewPeriod / 2
```
Comment (gambol99) [inline vault.go]: I guess an issue with this is, assuming a few errors or service unavailable, you can negative pretty quickly and start to eat cycles, perhaps minimum or some former of backoff

Comment (roboll) [reply vault.go]: This is true. 

Putting a backoff in makes sense I guess, but it is kind of weird to continue trying after the ttl of the secret is expired. 

Maybe we should just exit 1 when the renewPeriod gets below a minimum. What do you think?

File: vault.go (hunk 1 above)
Comment (roboll) [reply vault.go]: I don't think so - lookup self has always been around afaik, it just wasn't exposed the same way in the client wrapper. 

https://www.vaultproject.io/api/auth/token/index.html#lookup-a-token-self-

File: vault.go (hunk 2 above)
Comment (gambol99) [reply vault.go]: Hi @roboll, sorry for delay ... yes, agreed .. I'd rather exit 1 on a negative ttl and allow kubernetes to do the backoff (hopefully the user should have monitoring in place for the crashloopbackoff) and perhaps exchange the / 2 for a either a [backoff](https://github.com/jpillora/backoff) or static time.After? ... what you reckon?

**Comment by roboll** [comment]
@gambol99 I just pushed a change here to fail and exit if the renew period is less than 1s. I think that should be enough coverage for the fatal case where vault becomes totally unavailable.

File: vault.go [hunk 3]
```diff
@@ -481,6 +481,44 @@ func newVaultClient(opts *config) (*api.Client, error) {
...
+		}
+		glog.Infof("token ttl is %v", tokenttl)
+		renewPeriod := tokenttl / 2
+		go func() {
+			for {
+				glog.Infof("scheduling token renew in %v", renewPeriod)
+				<-time.After(renewPeriod)
+
+				glog.Infof("attempting token renew")
+				newtokeninfo, err := client.Auth().Token().RenewSelf(0)
```
Comment (gambol99) [inline vault.go]: quick question ... looking at [RenewSelf](https://github.com/hashicorp/vault/blob/master/api/auth_token.go#L121-L136) on error it's a nil .. should there not be a ```continue``` in the [error catch](https://github.com/UKHomeOffice/vault-sidekick/pull/61/files#diff-f95859fde2c2af3fd61c1b714a161d65R503) other ```newtokeninfo.TokenTTL()``` is gonna cause a nil pointer and then chunk the ```if renewPeriod < 1*time.Second {}``` at the near top of the loop

Comment (roboll) [reply vault.go]: Yes, good ideas. Changes made.

**Comment by roboll** [comment]
paging @gambol99

**Comment by gambol99** [comment]
LTGM ... much appreciated @roboll

**Comment by roboll** [comment]
@gambol99 thanks! would love to see a release with this - want me to set up a release pr or do you want to handle that yourself?

**Pull request merged**